test/extensions/filters/http/rate_limit_quota/integration_test.cc

```diff
@@ -412,6 +412,87 @@ TEST_P(RateLimitQuotaIntegrationTest, BasicFlowPeriodicalReport) {
   rlqs_stream_->sendGrpcMessage(rlqs_response2);
 }
 
+TEST_P(RateLimitQuotaIntegrationTest, MultiRequestWithTokenBucketThrottling) {
+  initializeConfig();
+  HttpIntegrationTest::initialize();
+  absl::flat_hash_map<std::string, std::string> custom_headers = {{"environment", "staging"},
+                                                                  {"group", "envoy"}};
+  int max_token = 1;
+  int tokens_per_fill = 30;
+  int fill_interval_sec = 60;
+  int fill_one_token_in_ms = fill_interval_sec / tokens_per_fill * 1000;
+  // First  request: allowed; fail-open, default no assignment policy.
+  // Second request: allowed; one token remaining, token bucket's max_token is 1.
+  // Third  request: allowed; token bucket has been refilled by advancing 2s.
+  // Fourth request: rejected; no token left and no token refilled.
+  // Fifth  request: allowed; token bucket has been refilled by advancing 2s.
+  // Sixth  request: rejected; no token left and no token refilled.
+  for (int i = 0; i < 6; ++i) {
```

Comment:
I think this test has built in race condition. It assumes that rate limit assignment sent by the fake RLQS services arrives before the second request is sent. However if the the fake upstream thread is slow, the second request may be processed before RLQS policy, and still use the default policy. This will allow the 4th request to succeed as there will be 1 more token in the bucket.

Reply:
You may want to wait for count of completed requests for the RLQS service cluster to be 1 before sending the second request.

Reply:
Per offline discussion, currently there no race condition here. 
But @yanavlasov has pointed out a good point: it will be a problem when TODO that checks no_assignment_behavior is implemented. Will check leverage stats counter check to address this.